collection-scripts/gather_machineconfig_ondisk

```diff
@@ -0,0 +1,34 @@
+#!/bin/bash 
+
+BASE_COLLECTION_PATH="must-gather"
+MACHINECONFIG_CONFIG_PATH=${OUT:-"${BASE_COLLECTION_PATH}/machine_config_ondisk"}
+
+mkdir -p "${MACHINECONFIG_CONFIG_PATH}"/
+
+# gather_machineconfig_ondisk_currentconfig collects the MCO's "on-disk" MachineConfig files
+function gather_machineconfig_ondisk {
+     
+    echo "INFO: Gathering on-disk MachineConfig from degraded nodes"
+    # we only really need the info from nodes that the MCO has marked as degraded 
+    MACHINECONFIG_DEGRADED_NODES=$(oc get node -o jsonpath='{.items[?(@.metadata.annotations.machineconfiguration\.openshift\.io/state=="Degraded")].metadata.name}')
+    for NODE in ${MACHINECONFIG_DEGRADED_NODES}; do
+         # use the existing MCD pod on the node to get the files
+         DAEMONPOD=$(oc get pods -n openshift-machine-config-operator --field-selector spec.nodeName=${NODE} --selector k8s-app=machine-config-daemon -o custom-columns=:metadata.name --no-headers)
+         # collect the Node's currentconfig that it believes it has applied 
+         timeout -v 3m oc cp -c machine-config-daemon openshift-machine-config-operator/"${DAEMONPOD}":rootfs/etc/machine-config-daemon/currentconfig "${MACHINECONFIG_CONFIG_PATH}"/"${NODE}"/currentconfig &
```

Comment:
Hmm, I am looking at the codepaths, and it seems like in the failure mode we care about here (the node had a drift from the bootstrap time config) we wouldn't have written this. We should be able to access this via the gathered MachineConfig in the general case. Is there a specific scenario/use case you were thinking of for this?

Reply:
also, if this fails (timeouts), does it affect the rest of the must-gather at all or would must-gather continue and ignore it?

Reply:
> Is there a specific scenario/use case you were thinking of for this?

There were multiple bugs in 1q this year that were like:

- customer's nodes are degraded and
- they deleted the rendered config while trying to fix the problem and
- the config the broken node applied caused issues but we don't know what it was

And in all those cases they "forcefiled" out and couldn't easily do root cause after the fact. ( And in my brain at least these were similar to the bootstrap cases in that they were "times when the node had a different state than the live cluster and we wanted to see the difference". ) 

I can take it out if it's contentious/the juice isn't worth the squeeze, not trying to be wasteful. I'm not *that* attached to it. I was just the one making the grocery list so I added the snacks I wanted  :smile: 

> also, if this fails (timeouts), does it affect the rest of the must-gather at all or would must-gather continue and ignore it? 

It will log that it failed but otherwise continue, the rest of the gather won't be impacted. If you want to try it, I have this one set to collect a file that doesn't exist from every node: 
```
oc adm must-gather --image quay.io/jkyros/must-gather:broken
```

you should get something like this: 
```
[must-gather-dk7s6] POD 2023-04-28T19:52:35.565600285Z INFO: Gathering on-disk MachineConfig from degraded nodes
[must-gather-dk7s6] POD 2023-04-28T19:52:35.906042203Z tar: rootfs/jkyros/nope: Cannot stat: No such file or directory
[must-gather-dk7s6] POD 2023-04-28T19:52:35.906042203Z tar: Exiting with failure status due to previous errors
[must-gather-dk7s6] POD 2023-04-28T19:52:36.001738289Z tar: rootfs/jkyros/nope: Cannot stat: No such file or directory
[must-gather-dk7s6] POD 2023-04-28T19:52:36.001738289Z tar: Exiting with failure status due to previous errors
[must-gather-dk7s6] POD 2023-04-28T19:52:36.285140754Z tar: rootfs/jkyros/nope: Cannot stat: No such file or directory
[must-gather-dk7s6] POD 2023-04-28T19:52:36.285140754Z tar: Exiting with failure status due to previous errors
[must-gather-dk7s6] POD 2023-04-28T19:52:36.383517740Z tar: rootfs/jkyros/nope: Cannot stat: No such file or directory
[must-gather-dk7s6] POD 2023-04-28T19:52:36.383517740Z tar: Exiting with failure status due to previous errors
[must-gather-dk7s6] POD 2023-04-28T19:52:36.468223813Z INFO: Waiting for on-disk MachineConfig collection to complete ...
[must-gather-dk7s6] POD 2023-04-28T19:52:36.526198943Z tar: rootfs/jkyros/nope: Cannot stat: No such file or directory
[must-gather-dk7s6] POD 2023-04-28T19:52:36.526198943Z tar: Exiting with failure status due to previous errors
[must-gather-dk7s6] POD 2023-04-28T19:52:36.647033091Z tar: rootfs/jkyros/nope: Cannot stat: No such file or directory
[must-gather-dk7s6] POD 2023-04-28T19:52:36.647033091Z tar: Exiting with failure status due to previous errors
[must-gather-dk7s6] POD 2023-04-28T19:52:36.649267872Z INFO: on-disk MachineConfig config collection complete.

```